<div style="text-align: center;">
    <h1>HAMOYE STAGE A TEST CODE SOLUTIONS</h1>
</div>


## STEP ONE: LOAD NECESSARY LIBRARIES
- Here, for the sake of this project, only the pandas libraies was used

In [1]:
import pandas as pd

## STEP TWO: PERFORM DATA WRANGLING/CLEANING OPERATIONS
1. Load the data
2. Preview the data
3. Check data type
4. Check for missing values
5. Check for duplicates
6. Check for inconsistent data
7. Explore basic statistics of the data


## **1. Load the data**

In [2]:
url="https://github.com/HamoyeHQ/HDSC-Introduction-to-Python-for-machine-learning/files/7768140/FoodBalanceSheets_E_Africa_NOFLAG.csv"
data=pd.read_csv(url,encoding="latin-1", on_bad_lines="skip")

## **2. Preview the data**
- Viewing the first five rows would give me a first-hand glimpse about the structure of the dataset and the information it contains

In [3]:
data.head()

Area Code     Area  Item Code         Item  Element Code  \
0          4  Algeria       2501   Population           511   
1          4  Algeria       2501   Population          5301   
2          4  Algeria       2901  Grand Total           664   
3          4  Algeria       2901  Grand Total           674   
4          4  Algeria       2901  Grand Total           684   

                                  Element             Unit     Y2014  \
0           Total Population - Both sexes     1000 persons  38924.00   
1                Domestic supply quantity      1000 tonnes      0.00   
2           Food supply (kcal/capita/day)  kcal/capita/day   3377.00   
3  Protein supply quantity (g/capita/day)     g/capita/day     94.90   
4      Fat supply quantity (g/capita/day)     g/capita/day     80.06   

      Y2015     Y2016     Y2017     Y2018  
0  39728.00  40551.00  41389.00  42228.00  
1      0.00      0.00      0.00      0.00  
2   3379.00   3372.00   3341.00   3322.00  
3     94.35     94.72     92.82     91.83  
4     79.36     77.40     80.19     77.28

## **3. Check data type**
- I would like to check if every column in the dataset contains a unique(the same) data types
- In order not to waste time by using the dtype() on each column, I will prefer to use a for-loop statement to perform this task across the whole dataset columns at once

In [4]:
# Create an empty DataFrame to store the results
data_type_df = pd.DataFrame(columns=['COLUMN NAMES', 'UNIQUE DATA TYPES'])

# Iterate through columns and store unique data types in the new DataFrame
for column_name in data.columns:
    unique_data_types = data[column_name].apply(type).unique()
    
    #data_type_df = data_type_df.append({'COLUMN NAME': column_name, 'UNIQUE DATA TYPES': unique_data_types}, ignore_index=True)
    data_type_df = pd.concat([data_type_df, pd.DataFrame({'COLUMN NAMES': [column_name], 'UNIQUE DATA TYPES': [unique_data_types]})],ignore_index=True)

# Display the resulting table
print(data_type_df)

    COLUMN NAMES  UNIQUE DATA TYPES
0      Area Code    [<class 'int'>]
1           Area    [<class 'str'>]
2      Item Code    [<class 'int'>]
3           Item    [<class 'str'>]
4   Element Code    [<class 'int'>]
5        Element    [<class 'str'>]
6           Unit    [<class 'str'>]
7          Y2014  [<class 'float'>]
8          Y2015  [<class 'float'>]
9          Y2016  [<class 'float'>]
10         Y2017  [<class 'float'>]
11         Y2018  [<class 'float'>]


### **INFERENCE:**
- Each column contains a unique data type, hence there would be no need for any type casting

## **4. Check for missing values**

In [5]:
data.isnull().sum()

Area Code          0
Area               0
Item Code          0
Item               0
Element Code       0
Element            0
Unit               0
Y2014           1589
Y2015           1548
Y2016           1535
Y2017           1506
Y2018           1436
dtype: int64

### **INFERENCE:**
- It appears that the last 5 columns (i.e. columns: Y2014, Y2015, Y2016, Y2017, Y2018) contains null values
- I would like to deep further by getting all the specific rows with missing values in the next code cell.

In [6]:
rows_with_missing_values = data.isnull().any(axis=1)
missing_data=data[rows_with_missing_values]
missing_data

Area Code      Area  Item Code                  Item  Element Code  \
46             4   Algeria       2805     Rice and products          5911   
61             4   Algeria       2513   Barley and products          5911   
88             4   Algeria       2515      Rye and products          5611   
100            4   Algeria       2516                  Oats          5911   
157            4   Algeria       2532  Cassava and products          5611   
...          ...       ...        ...                   ...           ...   
60715        181  Zimbabwe       2743                 Cream          5911   
60757        181  Zimbabwe       2949                  Eggs          5911   
60761        181  Zimbabwe       2949                  Eggs          5154   
60771        181  Zimbabwe       2744                  Eggs          5911   
60775        181  Zimbabwe       2744                  Eggs          5154   

                     Element         Unit  Y2014  Y2015  Y2016  Y2017  Y2018  
46           Export Quantity  1000 tonnes    NaN    0.0    0.0    0.0    0.0  
61           Export Quantity  1000 tonnes    NaN    0.0    0.0    NaN    0.0  
88           Import Quantity  1000 tonnes    NaN    NaN    NaN    NaN    0.0  
100          Export Quantity  1000 tonnes    NaN    NaN    0.0    0.0    0.0  
157          Import Quantity  1000 tonnes    0.0    NaN    0.0    0.0    NaN  
...                      ...          ...    ...    ...    ...    ...    ...  
60715        Export Quantity  1000 tonnes    0.0    NaN    0.0    0.0    NaN  
60757        Export Quantity  1000 tonnes    0.0    NaN    NaN    0.0    NaN  
60761  Other uses (non-food)  1000 tonnes    0.0    0.0    NaN    0.0    0.0  
60771        Export Quantity  1000 tonnes    0.0    NaN    NaN    0.0    NaN  
60775  Other uses (non-food)  1000 tonnes    0.0    0.0    NaN    0.0    0.0  

[3159 rows x 12 columns]

## **INFERENCE:**
- It appears that there are 3,159 unique rows with missing values
- Since, there were no specific instructions to fill the missing values, hence, I just intend to point this out.

## 5. Check for duplicates

In [7]:
duplicates = data.duplicated()
data[duplicates]


Empty DataFrame
Columns: [Area Code, Area, Item Code, Item, Element Code, Element, Unit, Y2014, Y2015, Y2016, Y2017, Y2018]
Index: []

### **INFERENCE:**
- It appears that there are no any duplicated records in the data set
- I can bet that I am a step close to certifying as READY! for analysis

## 6. Check for data inconsistency among the columns
- The dataset contains 4 columns with categorical values (Columns: Area, Item, Element, and Unit). Hence, it is important to check for value inconsitencies within each of these columns
- Some of the useful functions to use here are **value_count** and **nunique()**

In [8]:
#This would allow me to be able to view all the series, not allowing Python to shorten the display for me
pd.set_option('display.max_rows', None)

data['Element'].value_counts()

Domestic supply quantity                  5295
Import Quantity                           5139
Protein supply quantity (g/capita/day)    5023
Fat supply quantity (g/capita/day)        5023
Food supply (kcal/capita/day)             5014
Food                                      4941
Food supply quantity (kg/capita/yr)       4905
Residuals                                 4655
Export Quantity                           4403
Stock Variation                           4232
Production                                3881
Processing                                2010
Losses                                    2009
Other uses (non-food)                     1732
Feed                                      1319
Seed                                       762
Tourist consumption                        555
Total Population - Both sexes               45
Name: Element, dtype: int64

### **INFERENCE :**
- Upon checking, there seems not to be any inconsistency among the values within the categorical columns (The **AREA** column as a case study)
- To be better certify this conclusion, the metadata about the dataset would be helpful. For instance, the dataset metadata would contain information about the exact unique countries to expect in the dataset

## 7. Explore basic statistics of the data

In [9]:
data.describe()

Area Code     Item Code  Element Code          Y2014          Y2015  \
count  60943.000000  60943.000000  60943.000000   59354.000000   59395.000000   
mean     134.265576   2687.176706   3814.856456     134.196282     135.235966   
std       72.605709    146.055739   2212.007033    1567.663696    1603.403984   
min        4.000000   2501.000000    511.000000   -1796.000000   -3161.000000   
25%       74.000000   2562.000000    684.000000       0.000000       0.000000   
50%      136.000000   2630.000000   5142.000000       0.090000       0.080000   
75%      195.000000   2775.000000   5511.000000       8.340000       8.460000   
max      276.000000   2961.000000   5911.000000  176405.000000  181137.000000   

               Y2016          Y2017          Y2018  
count   59408.000000   59437.000000   59507.000000  
mean      136.555222     140.917765     143.758381  
std      1640.007194    1671.862359    1710.782658  
min     -3225.000000   -1582.000000   -3396.000000  
25%         0.000000       0.000000       0.000000  
50%         0.080000       0.100000       0.070000  
75%         8.430000       9.000000       9.000000  
max    185960.000000  190873.000000  195875.000000

### **INFERENCE:**
- Here, we can observe the basic stats of the numerical columns.
- For instance, the mean of the **Y2015** column is **135.24** (to two places of decimal)

## STEP THREE: PERFORM EXPLORATORY DATA ANLYSIS (EDA)
- **QUESTION1:** Which year had the least correlation with 'Element Code'
- **QUESTION3:** Select columns 'Y2017' and 'Area', Perform a groupby operation on Area. Which of these areas had the highest sum in Y2017?
- **QUESTION7:** What is the total number and percentage of missing data in Y2014 to 3 decimal places?
- **QUESTION8:** Select columns 'Y2017' and 'Area', Perform a groupby operation on Area. Which of these areas had the 7th lowest sum in Y2017?
- **QUESTION9:** What is the total Protein supply quantity in Madagascar in 2015?
- **QUESTION11:** What is the total sum of Wine produced in 2015 and 2018 respectively?(Perform a groupby sum aggregation on item)
- **QUESTION14:** What is the mean and standard deviation across the whole dataset for the year 2017 to 2 decimal places?
- **QUESTION16:** What is the total number of unique countries in the dataset?
- **QUESTION17:** Perform a groupby operation on 'Element'. What is the total number of the sum of Processing in 2017?
- **QUESTION19:** Perform a groupby operation on 'Element'. What year has the highest sum of stock variation?
- **QUESTION20:** How would you check for the number of rows and columns in a pandas DataFrame named df?

## **QUESTION 1**
- Which year had the least correlation with 'Element Code'

In [10]:
# Identify the 'Element Code' column
element_code_column = 'Element Code'

# Identify the year columns (assuming they are labeled as 'YearYYYY')
year_columns = [col for col in data.columns if col.startswith('Y')]

# Initialize variables to store minimum correlation and corresponding year
min_correlation = 1.0  # Set an initial high value
min_correlation_year = None

# Calculate correlations and find the year with the least correlation
for year_column in year_columns:
    correlation = data[element_code_column].corr(data[year_column])
    if abs(correlation) < abs(min_correlation):
        min_correlation = correlation
        min_correlation_year = year_column

# Print the year with the least correlation
print(f"The year with the least correlation to 'Element Code' is {min_correlation_year} with a correlation coefficient of {min_correlation:.2f}")

The year with the least correlation to 'Element Code' is Y2016 with a correlation coefficient of 0.02


### ANSWER: Y2016

## QUESTION 3
- Select columns 'Y2017' and 'Area', Perform a groupby operation on Area. Which of these areas had the highest sum in Y2017?

In [11]:
# Select columns 'Y2017' and 'Area'
selected_columns1 = data[['Area', 'Y2017']]

# Perform a groupby operation on 'Area' and calculate the sum for 'Y2017'
grouped_data1 = selected_columns1.groupby('Area')['Y2017'].sum().reset_index()

# Find the area with the highest sum in 'Y2017'
max_area = grouped_data1[grouped_data1['Y2017'] == grouped_data1['Y2017'].max()]

# Print the area with the highest sum in 'Y2017'
print("Area with the highest sum in Y2017:")
print(max_area)


Area with the highest sum in Y2017:
       Area       Y2017
34  Nigeria  1483268.23


### ANSWER: **Nigeria**
- Another way of simply getting the answer would by the method in the next code cell

In [12]:
grouped_data1.sort_values(by='Y2017', ascending=False)

Area       Y2017
34                      Nigeria  1483268.23
13                        Egypt   866379.92
40                 South Africa   517590.54
15                     Ethiopia   448683.76
30                      Morocco   388495.36
19                        Ghana   337599.06
0                       Algeria   325644.27
46  United Republic of Tanzania   322616.85
22                        Kenya   264660.66
41                        Sudan   239931.92
6                      Cameroon   232030.43
1                        Angola   229159.57
11                Côte d'Ivoire   224599.01
45                       Uganda   213950.38
26                       Malawi   181098.71
31                   Mozambique   161407.98
28                   Mauritania   156665.46
27                         Mali   149928.33
25                   Madagascar   131197.73
33                        Niger   126707.58
2                         Benin   124771.22
44                      Tunisia   124167.20
47                       Zambia   103223.77
4                  Burkina Faso   101855.07
20                       Guinea    98138.87
37                      Senegal    95681.15
48                     Zimbabwe    75919.34
35                       Rwanda    73663.69
8                          Chad    71594.68
39                 Sierra Leone    55311.33
14                     Eswatini    54343.33
29                    Mauritius    51114.83
43                         Togo    49841.88
10                        Congo    41181.68
7      Central African Republic    29937.00
32                      Namibia    29874.89
24                      Liberia    29342.20
17                        Gabon    27979.64
18                       Gambia    23154.18
12                     Djibouti    22729.91
3                      Botswana    22101.30
23                      Lesotho    21267.96
21                Guinea-Bissau    19102.77
5                    Cabo Verde    14650.74
36        Sao Tome and Principe    12662.63
38                   Seychelles      442.34
9                       Comoros       59.84
42               Sudan (former)        0.00
16                 Ethiopia PDR        0.00

### QUESTION 7
-  What is the total number and percentage of missing data in Y2014 to 3 decimal places?

In [13]:
total_rows = len(data)
missing_Y2014_count = data['Y2014'].isnull().sum()
percent_of_missing_Y2014_count = (missing_Y2014_count/total_rows) *100


# Print the results
print(f"Total number of missing data in Y2014: {missing_Y2014_count}")
print(f"Percentage of missing data in Y2014: {percent_of_missing_Y2014_count:.3f}%")

Total number of missing data in Y2014: 1589
Percentage of missing data in Y2014: 2.607%


### ANSWER
- Total number of missing data in Y2014: 1589
- Percentage of missing data in Y2014: 2.607%

## QUESTION 8
- Select columns 'Y2017' and 'Area' and perform a groupby operation on Area
- Which of these areas had the 7th lowest sum in Y2017?

In [15]:
#Very similar solution algorithm to QUESTION 3
# Select columns 'Y2017' and 'Area'
selected_columns2 = data[['Area', 'Y2017']]

# Perform a groupby operation on 'Area' and calculate the sum for 'Y2017'
grouped_data2 = selected_columns2.groupby('Area')['Y2017'].sum().reset_index()

# Find the area with the lowest sum in 'Y2017'
min_area = grouped_data2[grouped_data2['Y2017'] == grouped_data2['Y2017'].min()]

# Print the area with the lowest sum in 'Y2017'
#print("Area with the lowest sum in Y2017:")
#print(min_area)

grouped_data2.sort_values(by='Y2017').head(8)

Area     Y2017
42         Sudan (former)      0.00
16           Ethiopia PDR      0.00
9                 Comoros     59.84
38             Seychelles    442.34
36  Sao Tome and Principe  12662.63
5              Cabo Verde  14650.74
21          Guinea-Bissau  19102.77
23                Lesotho  21267.96

### ANSWER
- From the result above, **GUINEA-BISSAU** is the Area with the 7th lowest sum

## QUESTION 9
- What is the total Protein supply quantity in Madagascar in 2015?

In [17]:
# Filter the DataFrame to include only rows where 'Area' is 'Madagascar' and 'Year' is 2015
filtered_data = data[(data['Area'] == 'Madagascar')]

filtered_data

# Calculate the total Protein supply quantity for Madagascar in 2015
total_protein_supply = filtered_data[filtered_data['Element']=='Protein supply quantity (g/capita/day)'].sum()

# Print the result
print("The total Protein supply quantity in Madagascar in 2015 is: ")
print(f"\n{total_protein_supply}")


The total Protein supply quantity in Madagascar in 2015 is: 

Area Code                                                   14319
Area            MadagascarMadagascarMadagascarMadagascarMadaga...
Item Code                                                  299405
Item            Grand TotalVegetal ProductsAnimal ProductsCere...
Element Code                                                74814
Element         Protein supply quantity (g/capita/day)Protein ...
Unit            g/capita/dayg/capita/dayg/capita/dayg/capita/d...
Y2014                                                      177.79
Y2015                                                      173.05
Y2016                                                       173.4
Y2017                                                      175.75
Y2018                                                      176.64
dtype: object


### ANSWER
- From the result above, the total protein supply quantity for Madasgar in 2015 is 173.05

## QUESTION 11
- What is the total sum of Wine produced in 2015 and 2018 respectively?(Perform a groupby sum aggregation on item)

In [18]:
selected_columns3 = data[['Item', 'Y2015', 'Y2018']]
selected_columns3


grouped_data3 = selected_columns3.groupby('Item')['Y2015', 'Y2018'].sum().reset_index()
grouped_data3


wine_produced = grouped_data3[grouped_data3['Item'] == 'Wine']

wine_produced


C:\Users\msunmola\AppData\Local\Temp\ipykernel_14764\3617239110.py:5: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  grouped_data3 = selected_columns3.groupby('Item')['Y2015', 'Y2018'].sum().reset_index()


Item    Y2015    Y2018
117  Wine  4251.81  4039.32

### ANSWER
- The sum of wine produced in 2015 is 4251.81
- The sum of wine produced in 2018 is 4039.32

## QUESTION 14
- What is the mean and standard deviation across the whole dataset for the year 2017 to 2 decimal places?

In [19]:
data['Y2017'].describe().round(2)

count     59437.00
mean        140.92
std        1671.86
min       -1582.00
25%           0.00
50%           0.10
75%           9.00
max      190873.00
Name: Y2017, dtype: float64

### ANSWER
- The **mean** and **standard deviation** across the whole dataset for Y2017 are **140.92** and **1671.86** respectively

## QUESTION 16
- What is the total number of unique countries in the dataset?

In [20]:
data['Area'].nunique()

49

### ANSWER
- The total number of unique countries in the dataset is **49**

## QUESTION 17
- Perform a groupby operation on 'Element'
- What is the total number of the sum of Processing in 2017?

In [21]:
selected_columns4= data[['Item', 'Element', 'Y2017']]
selected_columns4


grouped_data4 =selected_columns4.groupby('Element')['Y2017'].sum()
grouped_data4


Element
Domestic supply quantity                  2088198.10
Export Quantity                            182338.80
Fat supply quantity (g/capita/day)          10253.84
Feed                                       223705.68
Food                                      1258888.28
Food supply (kcal/capita/day)              454681.00
Food supply quantity (kg/capita/yr)         48690.04
Import Quantity                            294559.09
Losses                                     160614.00
Other uses (non-food)                       91645.97
Processing                                 292836.00
Production                                2030056.89
Protein supply quantity (g/capita/day)      11842.45
Residuals                                   35500.00
Seed                                        24870.14
Stock Variation                             54316.91
Total Population - Both sexes             1112641.00
Tourist consumption                            91.00
Name: Y2017, dtype: float64

### ANSWER
- From the result above, the total sum of Processing is **292836.00**

## QUESTION 19
- Perform a groupby operation on 'Element'
- What year has the highest sum of stock variation?

In [22]:
selected_column5= data[['Element', 'Y2014','Y2015', 'Y2016', 'Y2017', 'Y2018']]



grouped_data5 =selected_column5.groupby('Element')['Y2014','Y2015', 'Y2016', 'Y2017', 'Y2018'].sum().max()
grouped_data5

C:\Users\msunmola\AppData\Local\Temp\ipykernel_14764\3659313714.py:5: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  grouped_data5 =selected_column5.groupby('Element')['Y2014','Y2015', 'Y2016', 'Y2017', 'Y2018'].sum().max()


Y2014    1996716.35
Y2015    2021493.55
Y2016    2044842.70
Y2017    2088198.10
Y2018    2161192.10
dtype: float64

### ANSWER
- 2018

## QUESTION 20
- How would you check for the number of rows and columns in a pandas DataFrame named df?

In [23]:
data.shape

(60943, 12)

### ANSWER
- The shape method is used to check for the number of rows and columns.
- The result above shows the dataset is having **60943** rows and **12** columns